### PR TITLE
Adding Llama to TorchAO

### PR DESCRIPTION
**Pull request by HDCharles**
Summary: This PR adds funcitonality for stable eval/benchmarking of llama models within the torchao codebase.

the model stuff is in torchao/_models/llama with eval being moved to _models/_eval.py

model.py was updated to work with llama3 and edited such that a new configuration variable use_index_put_for_kv_cache can be used to set the kv_cache update method. This allows it to work for both benchmarking as well as GPTQ.

eval.py was updated to work with llama3, though the result runs, the numbers do not seem to be correct and need further investigation. eval does work for llama2 though

tokenizer.py was added for llama3 and tiktoken support which was copied from gpt-fast

new benchmarking code in torchao/_models/llama i.e. benchmarks.sh generate.py and benchmark_results.txt contain a stable way to benchmark models using torchao APIs. These scripts were adapted from gpt-fast

The quantization README was updated with our current benchmark and evaluation numbers (eval only for llama2)
    
The test_quant_api.py tests were updated to use the new tokenizer code and an example of the issue with llama3 can be seen in the test_eval_wrapper_llama3 test.

Test Plan:
    
python test/quantization/test_quant_api.py
sh torchao/_models/llama/benchmarks.sh
    
Reviewers:
    
Subscribers:
    
Tasks:
    
Tags:

### CODE REVIEW

File: test/quantization/test_quant_api.py [hunk 1]
```diff
@@ -282,21 +284,22 @@ def test_8da4w_quantizer_eval(self):
...
         )
 
-    @unittest.skip("skipping until we get checkpoints for gpt-fast")
+    # @unittest.skip("skipping until we get checkpoints for gpt-fast")
     def test_gptq_quantizer_int4wo(self):
         from torchao.quantization.GPTQ import Int4WeightOnlyGPTQQuantizer, InputRecorder, TransformerEvalWrapper
         precision = torch.bfloat16
         device = "cuda"
-        checkpoint_path = Path("../gpt-fast/checkpoints/meta-llama/Llama-2-7b-chat-hf/model.pth")
+        checkpoint_path = Path("/home/cdhernandez/local/gpt-fast/checkpoints/meta-llama/Llama-2-7b-chat-hf/model.pth")
```
Comment (andrewor14) [inline test/quantization/test_quant_api.py:292]: probably want to revert these?

File: test/quantization/test_quant_api.py [hunk 2]
```diff
@@ -43,8 +43,8 @@
     TORCH_VERSION_AFTER_2_4,
 )
 from pathlib import Path
-from sentencepiece import SentencePieceProcessor
-from model import Transformer, prepare_inputs_for_model
+from torchao.models.llama.tokenizer import get_tokenizer
+from torchao.models.llama.model import Transformer, prepare_inputs_for_model
```
Comment (andrewor14) [inline test/quantization/test_quant_api.py:47]: should we put these under `torchao._models` for now so people don't start using our models? @msaroufim  @cpuhrsch

File: torchao/models/llama/tokenizer.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
+import os
```
Comment (andrewor14) [inline torchao/models/llama/tokenizer.py:1]: I guess these are taken from gpt-fast. Can we add a comment somewhere to mention that?

**Review by andrewor14 — APPROVED** [review]
Thanks! Were you able to replicate the current non-GPTQ tests on llama3? Would be good to add them here or in a future PR

File: test/quantization/test_quant_api.py (hunk 2 above)
Comment (msaroufim) [reply test/quantization/test_quant_api.py:47]: Yeah that sounds good

File: torchao/models/llama/tokenizer.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            mergeable_ranks=mergeable_ranks,
+            special_tokens=self.special_tokens,
+        )
+        # BOS / EOS token IDs
+        self._bos_id: int = self.special_tokens["<|begin_of_text|>"]
+        self._eos_id: int = self.special_tokens["<|end_of_text|>"]
+
+    def encode(self, text):
+        return self.model.encode(text)
+
```
Comment (sanchitintel) [inline torchao/models/llama/tokenizer.py:87]: Hi @HDCharles, it'd be great if torchtune could also use `TikTokenWrapper`, because right now, if one were to use LLaMA3's own tokenizer with torchao's GPTQ, then one would run into a problem - [_eval.py uses only one argument with encode](https://github.com/pytorch/ao/blob/main/torchao/_eval.py#L97), but [TikToken also uses two arguments that don't have any default values](https://github.com/pytorch/torchtune/blob/main/torchtune/modules/tokenizers/_tiktoken.py#L230).

Since torchao is a dependency of torchtune, how about also adding this wrapper at https://github.com/pytorch/torchtune/blob/main/torchtune/models/llama3/_model_builders.py#L69, once this PR is merged? That'd prevent duplicating this code in torchtune.

Thanks!

File: test/quantization/test_quant_api.py (hunk 2 above)
Comment (sanchitintel) [reply test/quantization/test_quant_api.py:47]: Would torchtune users not be able to use them? :(

torchtune users can't currently [use ](https://pytorch.org/torchtune/stable/tutorials/llama3.html#faster-generation-via-quantization)torchao's GPTQ with a (fine-tuned) LLaMA3 because of [tokenization issues mentioned here](https://github.com/pytorch/ao/pull/276/files#r1619492654).

File: torchao/models/llama/tokenizer.py (hunk 4 above)
Comment (msaroufim) [reply torchao/models/llama/tokenizer.py:87]: So this model is not intended to be downloaded by other projects. The idea is we need to some fast way to run performance benchmarks and evaluations in the repo so this should be very much treated like a test utility.

Granted some of these annoyances are stemming from us not maintaining a standalone performant llama fork like gpt-fast that's pip installable and perhaps we should just do that - @kartikayk might have some more thoughts

File: test/quantization/test_quant_api.py (hunk 2 above)
Comment (HDCharles) [reply test/quantization/test_quant_api.py:47]: i'll edit GPTQ to get tiktoken to work in the next PR, this one is mostly about llama2 functionality.

Comment (sanchitintel) [reply test/quantization/test_quant_api.py:47]: Thanks in advance! :)

**Comment by HDCharles** [comment]
> 

no, i get weird eval results for llama3, still need to figure that out, teh benchmarking seems to work though

File: test/quantization/test_quant_api.py (hunk 2 above)
Comment (HDCharles) [reply test/quantization/test_quant_api.py]: tiktoken works with gptq but something is busted as far as llama3 eval goes, not sure what the issue is, the accuracy is just really bad
